src/pproc/common/io.py

```diff
@@ -397,12 +397,28 @@ def target_from_location(
 class GribMetadata(eccodes.Message):
     def __init__(self, handle, headers_only: bool = False):
         new_handle = eccodes.codes_clone(handle, headers_only=headers_only)
+        self.extra = {"bitsPerValue": eccodes.codes_get(handle, "bitsPerValue", int)}
         super().__init__(new_handle)
 
     def __getstate__(self) -> dict:
-        ret = {"_handle": self.get_buffer()}
+        ret = {"_handle": self.get_buffer(), "extra": self.extra}
         return ret
 
     def __setstate__(self, state: dict):
         state["_handle"] = eccodes.MemoryReader(state["_handle"])._next_handle()
         self.__dict__.update(state)
+
+    def set(self, *args, check_values: bool = True):
+        super().set(*args, check_values=check_values)
```

Comment:
Shouldn't we remove the keys from `self.extra` from the args first?

Reply:
Where possible, I changed it so that we only set the metadata in `write_grib` as this also allows all the values required to be checked in the set. However, there are some cases were this isn't possible and we need intermediate sets. I'd be happy to discuss this further if you think it would be better not set `bitsPerValue` here and have an additional flag/different method to do the final set with the `self.extra` keys.

Reply:
No worries! Happy to merge as is if that works as intended